src/test.js

```diff
@@ -297,7 +297,10 @@ Test.prototype = {
 
 		// After emitting the js-reporters event we cleanup the assertion data to
 		// avoid leaking it. It is not used by the legacy testDone callbacks.
-		emit( "testEnd", this.testReport.end( true ) );
+		const promise = emitWithPromise( "testEnd", this.testReport.end( true ) );
+		if ( promise ) {
+			this.resolvePromise( promise );
```

Comment:
In the event we have a Promise, we'll need to block execution of the rest of this block. Otherwise, we're likely to run into ordering issues. For example, the `testEnd` hook may still be running but `QUnit.config.current` may be `undefined` or `suiteEnd` may execute.